File: app/src/main/java/com/saud/taskstrip/sync/SyncNoteDocument.kt
package com.saud.taskstrip.sync

import org.json.JSONArray
import org.json.JSONObject

/** One synced text, as it travels.
 *
 * `id` is a UUID string minted once and never regenerated — it is the only thing that lets two
 * devices agree they are looking at the same note rather than at two notes that happen to say the
 * same words. Neither app's own database key is it: this one's is an autoincrementing Long and the
 * Mac's is a SwiftData object, and neither means anything on the other machine. */
data class SyncNoteRecord(
    val id: String,
    val title: String = "",
    val text: String = "",
    val updatedAt: Long = System.currentTimeMillis(),
    /** A deleted note stays in the document as a tombstone. Dropping the row instead would make a
     * delete indistinguishable from "the other device hasn't heard of this yet", and the note
     * would come back from the dead on the next sync. */
    val isDeleted: Boolean = false
) {
    /** What to call it in a list. Mirrors SyncNoteRecord.displayTitle on the Mac exactly. */
    val displayTitle: String
        get() {
            val trimmedTitle = title.trim()
            if (trimmedTitle.isNotEmpty()) return trimmedTitle
            val line = text.lineSequence().firstOrNull { it.isNotBlank() }?.trim().orEmpty()
            if (line.isEmpty()) return "Untitled"
            return if (line.length <= 80) line else line.take(80)
        }
}

/** The document both apps read and write: one JSON file in the Drive folder they already share.
 *
 * This file is a contract with SyncNoteDocument.swift. Every rule below has to hold identically
 * there, because the phone and the Mac merge the same document independently and have to reach the
 * same answer without talking to each other.
 *
 * Deliberately one file rather than one per note. What one file costs is a same-millisecond edit
 * of the same note from both devices, which loses one side — see [winner]. What it buys is one
 * download, one upload, and one atomic replacement. */
object SyncNoteDocument {
    const val FILE_NAME = "sync_notes.json"
    const val MIME_TYPE = "application/json"
    const val VERSION = 1

    // ---- Reading and writing ----

    fun toJson(notes: List<SyncNoteRecord>): String {
        val array = JSONArray()
        sorted(notes).forEach { note ->
            array.put(
                JSONObject().apply {
                    put("id", note.id)
                    put("title", note.title)
                    put("text", note.text)
                    put("updatedAt", note.updatedAt)
                    put("deleted", note.isDeleted)
                }
            )
        }
        return JSONObject().apply {
            put("version", VERSION)
            put("notes", array)
        }.toString(2)
    }

    /** Anything unreadable is no notes rather than an error. A corrupt or half-written document in
     * Drive must not be able to wipe what's on this device — the merge treats "no remote" as
     * "nothing to add", and the next upload writes a good document over it. */
    fun fromJson(json: String): List<SyncNoteRecord> = runCatching {
        val array = JSONObject(json).getJSONArray("notes")
        (0 until array.length()).mapNotNull { index ->
            val obj = array.optJSONObject(index) ?: return@mapNotNull null
            val id = obj.optString("id")
            // An entry with no id can't be merged against anything, so it's dropped rather than
            // given a fresh id and duplicated on every sync.
            if (id.isNullOrEmpty()) return@mapNotNull null
            SyncNoteRecord(
                id = id,
                title = obj.optString("title", ""),
                text = obj.optString("text", ""),
                updatedAt = obj.optLong("updatedAt", 0L),
                isDeleted = obj.optBoolean("deleted", false)
            )
        }
    }.getOrDefault(emptyList())

    // ---- Merging ----

    /** Both sides' notes, one winner per id.
     *
     * Order-independent on purpose: `merge(local, remote)` and `merge(remote, local)` produce the
     * same thing, so the phone and the Mac reach the same answer without having to agree on who
     * went first. That is what [winner]'s tie-breaks are for. */
    fun merge(local: List<SyncNoteRecord>, remote: List<SyncNoteRecord>): List<SyncNoteRecord> {
        val byId = LinkedHashMap<String, SyncNoteRecord>()
        (local + remote).forEach { note ->
            val existing = byId[note.id]
            byId[note.id] = if (existing == null) note else winner(existing, note)
        }
        return sorted(byId.values.toList())
    }

    /** Newer wins. Then a delete wins over a live note, because a delete is a decision and a stale
     * edit is not. Then the greater text, purely so that two devices holding genuinely different
     * text stamped at the same millisecond still pick the *same* side and stop disagreeing. */
    fun winner(a: SyncNoteRecord, b: SyncNoteRecord): SyncNoteRecord {
        if (a.updatedAt != b.updatedAt) return if (a.updatedAt > b.updatedAt) a else b
        if (a.isDeleted != b.isDeleted) return if (a.isDeleted) a else b
        if (a.text != b.text) return if (isGreater(a.text, b.text)) a else b
        if (a.title != b.title) return if (isGreater(a.title, b.title)) a else b
        return a
    }

    /** Compares UTF-8 bytes, not strings.
     *
     * Kotlin's `>` on a String orders UTF-16 code units; Swift's orders grapheme clusters under
     * canonical equivalence. For ASCII they agree and for Arabic they need not — and a tie-break
     * the two platforms disagree about is worse than no tie-break at all, because the devices
     * would hand each other opposite answers forever. Bytes are the one ordering both can compute
     * the same way without either having to imitate the other's string semantics. */
    fun isGreater(a: String, b: String): Boolean {
        val x = a.toByteArray(Charsets.UTF_8)
        val y = b.toByteArray(Charsets.UTF_8)
        val shared = minOf(x.size, y.size)
        for (i in 0 until shared) {
            val left = x[i].toInt() and 0xFF
            val right = y[i].toInt() and 0xFF
            if (left != right) return left > right
        }
        return x.size > y.size
    }

    /** Newest first, and by id where the times match, so the file's bytes don't churn between
     * syncs that changed nothing. */
    fun sorted(notes: List<SyncNoteRecord>): List<SyncNoteRecord> =
        notes.sortedWith(compareByDescending<SyncNoteRecord> { it.updatedAt }.thenBy { it.id })

    /** What a person should see: tombstones are bookkeeping, not notes. */
    fun visible(notes: List<SyncNoteRecord>): List<SyncNoteRecord> =
        sorted(notes.filter { !it.isDeleted })
}

File: macos/TaskStrips/Sources/Core/Sync/SyncNoteDocument.swift
import Foundation

/// One synced text, as it travels.
///
/// `id` is a UUID string minted once and never regenerated — it is the only thing that lets two
/// devices agree they are looking at the same note rather than at two notes that happen to say the
/// same words. A row's local database key is never it: Android's is an autoincrementing Int and
/// the Mac's is a SwiftData object, and neither means anything on the other machine.
struct SyncNoteRecord: Identifiable, Equatable {
    var id: String
    /// The whole note. A synced note is one text with no title of its own.
    var text: String
    var updatedAt: Date
    /// A deleted note stays in the document as a tombstone. Dropping the row instead would make a
    /// delete indistinguishable from "the other device hasn't heard of this yet", and the note
    /// would come back from the dead on the next sync.
    var isDeleted: Bool

    /// What to call it in a list: its first non-blank line, the same rule the quick-notes
    /// scratchpad uses to name a strip it promotes.
    var displayTitle: String {
        let firstLine = text.split(separator: "\n").first { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
        let line = firstLine.map(String.init)?.trimmingCharacters(in: .whitespaces) ?? ""
        if line.isEmpty { return "Untitled" }
        return line.count <= 80 ? line : String(line.prefix(80))
    }
}

/// The document both apps read and write: one JSON file in the Drive folder they already share.
///
/// Deliberately one file rather than one per note. Two devices editing different notes at the same
/// time is the common case and a per-note file would handle it slightly better; but a single file
/// is one download, one upload and one atomic replacement, and the merge below already keeps both
/// devices' edits to *different* notes. What one file costs is a same-second edit of the *same*
/// note from both devices, which loses one side — see `winner`.
enum SyncNoteDocument {
    static let fileName = "sync_notes.json"
    static let mimeType = "application/json"

    /// 2 dropped the per-note `title` in favour of one text whose first line names it.
    ///
    /// Nothing reads this number to decide how to parse — a v1 document parses correctly under
    /// these rules, because `notes(from:)` still folds a title it finds into the text, and a v2
    /// document parses correctly under v1's rules too, because a missing title simply falls back
    /// to the first line there as well. It is written so the file says which app wrote it.
    static let version = 2

    /// A note written before the title was dropped, as one text.
    ///
    /// The title becomes the first line, which is the position the name is now read from — so a
    /// note called "Groceries" is still called "Groceries" afterwards. Must stay byte-identical to
    /// `foldLegacyTitle` in SyncNoteDocument.kt: the two devices fold independently and have to
    /// land on the same string, or the merge sees two different texts and picks a winner forever.
    static func foldLegacyTitle(_ title: String, _ text: String) -> String {
        if title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return text }
        if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return title }
        return title + "\n" + text
    }

    // MARK: - Reading and writing

    /// Milliseconds since the epoch, because that is what Android writes everywhere else in this
    /// app and the two have to read each other's numbers without a second thought.
    static func data(for notes: [SyncNoteRecord]) throws -> Data {
        let objects: [[String: Any]] = sorted(notes).map { note in
            [
                "id": note.id,
                "text": note.text,
                "updatedAt": Int(note.updatedAt.timeIntervalSince1970 * 1000),
                "deleted": note.isDeleted,
            ]
        }
        return try JSONSerialization.data(
            withJSONObject: ["version": version, "notes": objects],
            options: [.prettyPrinted, .sortedKeys]
        )
    }

    /// Anything unreadable is no notes rather than an error. A corrupt or half-written document in
    /// Drive must not be able to wipe what's on this machine — the merge treats "no remote" as
    /// "nothing to add", and the next upload writes a good document over it.
    static func notes(from data: Data) -> [SyncNoteRecord] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let array = root["notes"] as? [[String: Any]]
        else { return [] }

        return array.compactMap { object in
            guard let id = object["id"] as? String, !id.isEmpty else { return nil }
            let millis = (object["updatedAt"] as? NSNumber)?.doubleValue ?? 0
            return SyncNoteRecord(
                id: id,
                text: foldLegacyTitle(
                    object["title"] as? String ?? "",
                    object["text"] as? String ?? ""
                ),
                updatedAt: Date(timeIntervalSince1970: millis / 1000),
                isDeleted: object["deleted"] as? Bool ?? false
            )
        }
    }

    // MARK: - Merging

    /// Both sides' notes, one winner per id.
    ///
    /// Order-independent on purpose: `merge(local, remote)` and `merge(remote, local)` produce the
    /// same thing, so the phone and the Mac reach the same answer without having to agree on who
    /// went first. That is what `winner`'s tie-breaks are for.
    static func merge(local: [SyncNoteRecord], remote: [SyncNoteRecord]) -> [SyncNoteRecord] {
        var byID: [String: SyncNoteRecord] = [:]
        for note in local + remote {
            byID[note.id] = byID[note.id].map { winner($0, note) } ?? note
        }
        return sorted(Array(byID.values))
    }

    /// Newer wins. Then a delete wins over a live note, because a delete is a decision and a stale
    /// edit is not. Then the greater text, purely so that two devices holding genuinely different
    /// text stamped at the same millisecond still pick the *same* side and stop disagreeing.
    static func winner(_ a: SyncNoteRecord, _ b: SyncNoteRecord) -> SyncNoteRecord {
        if a.updatedAt != b.updatedAt { return a.updatedAt > b.updatedAt ? a : b }
        if a.isDeleted != b.isDeleted { return a.isDeleted ? a : b }
        if a.text != b.text { return isGreater(a.text, b.text) ? a : b }
        return a
    }

    /// Compares UTF-8 bytes, not Swift strings.
    ///
    /// `>` on a Swift String orders grapheme clusters under canonical equivalence; Kotlin's orders
    /// UTF-16 code units. For ASCII they agree and for Arabic they need not — and a tie-break the
    /// two platforms disagree about is worse than no tie-break at all, because the devices would
    /// hand each other opposite answers forever. Bytes are the one ordering both can compute the
    /// same way without either having to imitate the other's string semantics.
    static func isGreater(_ a: String, _ b: String) -> Bool {
        Array(b.utf8).lexicographicallyPrecedes(Array(a.utf8))
    }

    /// Newest first, and by id where the times match, so the file's bytes don't churn between
    /// syncs that changed nothing.
    static func sorted(_ notes: [SyncNoteRecord]) -> [SyncNoteRecord] {
        notes.sorted {
            $0.updatedAt == $1.updatedAt ? $0.id < $1.id : $0.updatedAt > $1.updatedAt
        }
    }

    /// What a person should see: tombstones are bookkeeping, not notes.
    static func visible(_ notes: [SyncNoteRecord]) -> [SyncNoteRecord] {
        sorted(notes.filter { !$0.isDeleted })
    }
}
